src/attr.rs

```diff
@@ -0,0 +1,64 @@
+use std::fmt;
+use std::borrow::{Borrow, Cow};
+
+use escape::Escaped;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// All underscores in the attribute name will be `Display`-ed as hyphens
+pub struct Attribute<'a> {
+    pub name: Cow<'a, str>,
+    pub value: Cow<'a, str>,
+}
+
+impl<'a> Attribute<'a> {
+    pub fn new<N, V>(name: N, value: V) -> Attribute<'a>
+        where N: Into<Cow<'a, str>>,
+              V: Into<Cow<'a, str>>
+    {
+        Attribute {
+            name: name.into(),
+            value: value.into(),
+        }
+    }
+}
+
+impl<'a> fmt::Display for Attribute<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let name: &str = self.name.borrow();
+        write!(f, "{}=\"{}\"", hyphenate(name.into()), Escaped(&self.value))
+    }
+}
+
+#[derive(Clone, Debug)]
+/// Use `attr_set!` macro instead of manual construction
+pub struct AttributeSet<'a>(pub Cow<'a, [Attribute<'a>]>);
+
+impl<'a> fmt::Display for AttributeSet<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for attr in self.0.iter() {
+            try!(write!(f, " {}", attr));
+        }
+        Ok(())
+    }
+}
+
+fn hyphenate<'a>(s: Cow<'a, str>) -> Cow<'a, str> {
+    let mut owned: Option<String> = None;
+    {
+        let bs: &str = s.borrow();
+        for (i, c) in bs.chars().enumerate() {
+            if c == '_' {
+                let mut os = owned.unwrap_or({
+                    let mut os = String::with_capacity(bs.len());
+                    os.extend((&bs[..i]).chars());
+                    os
+                });
+                os.push('-');
+                owned = Some(os);
+            } else if let Some(ref mut os) = owned {
+                os.push(c);
+            }
+        }
+    }
+    owned.map(|os| os.into()).unwrap_or(s)
```

Comment:
I don't think this will be a real performance issue, something simple would probably be fine. 

```rust
if s.contains('_') {
  Cow::Owned(s.replace('_', '-'))
} else {
  s
}
```

In fact, I'd almost be hopeful that the optimiser could basically run the `contains` and choose the branch at compile time.